File: Exp5A_GoBackN/GoBackN.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void transmission(int N, int tf) {
    int i = 1;
    int tt = 0;
    srand(time(NULL));
    while (i <= tf) {
        int z = 0;
        for (int k = i; k < i + N && k <= tf; k++) {
            printf("Sending Frame %d...\n", k);
            tt++;
        }
        for (int k = i; k < i + N && k <= tf; k++) {
            int f = rand() % 2;
            if (!f) {
                printf("Acknowledgment for Frame %d...\n", k);
                z++;
            } else {
                printf("Timeout!! Frame Number: %d Not Received\n", k);
                printf("Retransmitting Window...\n");
                break;
            }
        }
        printf("\n");
        i += z;
    }
    printf("Total number of frames which were sent and resent are: %d\n", tt);
}

int main() {
    int tf, N;
    printf("Enter the Total number of frames: ");
    scanf("%d", &tf);
    printf("Enter the Window Size: ");
    scanf("%d", &N);
    transmission(N, tf);
    return 0;
}

File: Exp5A_GoBackN/GoBackN.py
import random
import time

def transmission(N, tf):
    i = 1
    tt = 0
    while i <= tf:
        z = 0
        for k in range(i, min(i + N, tf + 1)):
            print("Sending Frame", k, "...")
            tt += 1
        for k in range(i, min(i + N, tf + 1)):
            f = random.randint(0, 1)
            if not f:
                print("Acknowledgment for Frame", k, "...")
                z += 1
            else:
                print("Timeout!! Frame Number:", k, "Not Received")
                print("Retransmitting Window...")
                break
        print()
        i += z
    return tt

def main():
    tf = int(input("Enter the Total number of frames: "))
    N = int(input("Enter the Window Size: "))
    random.seed(time.time())
    tt = transmission(N, tf)
    print("Total number of frames which were sent and resent are:", tt)

if __name__ == "__main__":
    main()
